modules/telemetry/json_packets.py

```diff
@@ -185,3 +188,167 @@ def parse_mission_file(mission_file: Path) -> MissionEntry:
+
+
+@dataclass
+class TelemetryDataPacketBlock:
+    """A generic block object to store information for telemetry data
+    All stored values must be updated at once!"""
+
+    mission_time: list[int] = field(default_factory=list[int])
+    stored_values: dict[str, list[int]] = field(default_factory=dict)
+
+    def __init__(self, keys: list[str]):
+        """Initializes a TelemetryDataPacketBlock"""
+        self.mission_time = []
+        self.stored_values = dict.fromkeys(keys, [])
+        logger.debug(self)
+
+    def update(self, data: dict[str, int], buffer_size: int) -> None:
+        """Updates the stored values with the given data
+        :param data: Dictionary of data to update containing mission_time and stored_values squashed
+        :param buffer_size: Size of the telemetry buffer to store"""
+        # Ensure you are not half updating the packet
+        # As this can cause the arrays to become out of sync and meaningless.
+        if dict(self).keys() != data.keys():
+            logger.error("Block must be updated using a full set of values at the same time!")
+            logger.debug(f"Tried updating {dict(self).keys()} using {data.keys()}!")
+            return
+
+        # Updates stored values with new values
+        for key in data.keys():
+            if key == "mission_time":
+                self.mission_time.append(data["mission_time"])
+            else:
+                self.stored_values[key].append(data[key])
+
+        # Ensure buffer sizes are kept
+        # Note: This uses a while loop incase buffer size shrinks drastically
+        while len(self.mission_time) > buffer_size:
+            self.mission_time.pop(0)
+            for key in self.stored_values.keys():
+                self.stored_values[key].pop(0)
+
+    def clear(self) -> None:
+        """Clears all stored values"""
+        self.mission_time = []
+        self.stored_values = dict.fromkeys(self.stored_values.keys(), [])
+
+    def __str__(self):
+        """Returns a string representation of the TelemetryDataPacketBlock"""
+        return f"{self.__class__.__name__} -> time: {self.mission_time} ms, {self.stored_values}"
+
+    def __iter__(self):
+        """Returns an interator containing all the stored values"""
+        yield "mission_time", self.mission_time
+        for key in self.stored_values.keys():
+            yield key, self.stored_values[key]
+
+
+@dataclass
+class TelemetryData:
+    """Contains the output specification for the telemetry data block"""
+
+    # Configuration
+    buffer_size: int = 20
+    decoder: list[dict[int, dict[str, str]]] = field(default_factory=list)
+
+    # Current data storage
+    last_mission_time: int = -1
+    output_blocks: dict[str, TelemetryDataPacketBlock] = field(default_factory=dict)
+    update_buffer: dict[str, dict[str, int | None]] = field(default_factory=dict)
+
+    def __init__(self, telemetry_buffer_size: int = 20):
```

Comment:
Once again; data class should not have defined `__init__` methods. I would make this a regular class.